bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/RestManager.java

```diff
@@ -0,0 +1,112 @@
+
+/**
+ * Base class for all various rest managers
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public abstract class RestManager {
+    private static final UriBuilder API_BASE_BUILDER = UriBuilder.fromUri(URL_API);
+    private static final UriBuilder APP_URI_BUILDER = UriBuilder.fromUri(URL_APP).path(PATH_API);
+    private static final UriBuilder API_URI_BUILDER = getApiBaseBuilder().path(PATH_API);
+
+    private final Set<Scope> requiredScopes;
+    private final ApiBridgeHandler apiBridge;
+
+    public RestManager(ApiBridgeHandler apiBridge, FeatureArea features) {
+        this.requiredScopes = features.scopes;
+        this.apiBridge = apiBridge;
+    }
+
+    protected <T extends ApiResponse<?>> T get(UriBuilder uriBuilder, Class<T> clazz) throws NetatmoException {
+        return executeUri(uriBuilder, HttpMethod.GET, clazz, null, null);
+    }
+
+    protected <T extends ApiResponse<?>> T post(UriBuilder uriBuilder, Class<T> clazz, @Nullable String payload,
+            @Nullable String contentType) throws NetatmoException {
+        return executeUri(uriBuilder, HttpMethod.POST, clazz, payload, contentType);
+    }
+
+    protected <T> T post(URI uri, Class<T> clazz, Map<String, String> entries) throws NetatmoException {
+        return apiBridge.executeUri(uri, POST, clazz, toRequest(entries),
+                "application/x-www-form-urlencoded;charset=UTF-8", 3);
+    }
+
+    private <T extends ApiResponse<?>> T executeUri(UriBuilder uriBuilder, HttpMethod method, Class<T> clazz,
+            @Nullable String payload, @Nullable String contentType) throws NetatmoException {
+        URI uri = uriBuilder.build();
+        T response = apiBridge.executeUri(uri, method, clazz, payload, contentType, 3);
+        if (response instanceof ApiResponse.Ok && ((ApiResponse.Ok) response).failed()) {
+            throw new NetatmoException("Command failed : %s for uri : %s", response.getStatus(), uri.toString());
+        }
+        return response;
+    }
+
+    private static UriBuilder appendParams(UriBuilder builder, @Nullable Object... params) {
+        if (params.length % 2 != 0) {
+            throw new IllegalArgumentException("appendParams : params count must be even");
+        }
+        for (int i = 0; i < params.length; i += 2) {
+            Object query = params[i];
+            if (query instanceof String) {
+                Object param = params[i + 1];
+                if (param != null) {
+                    builder.queryParam((String) query, param);
+                }
+            } else {
+                throw new IllegalArgumentException("appendParams : even parameters must be Strings");
+            }
+        }
+        return builder;
+    }
+
+    protected static UriBuilder getApiBaseBuilder() {
+        return API_BASE_BUILDER.clone();
+    }
+
+    public static UriBuilder getApiUriBuilder(String path, @Nullable Object... params) {
```

Comment:
IllegalArgumentException is not caught, neither here nor in the calling methods.
I would suggest to catch IllegalArgumentException and then throw a NetatmoException 
WDYT ?

Reply:
IMHO it should not be caught. `IllegalArgumentException` is a `RuntimeException` (thus unchecked) and designed to help detecting bugs in the code. The appropriate counter-measure is not to catch it, but to avoid having it thrown by calling with appropriate arguments.

Reply:
It could potentially lead to periodic jobs being no more rescheduled (I have not checked if this is the case here).
But if the proper way is to no not catch such exception, I close my comment.

Reply:
It's similar to avoiding division by zero instead of catching `ArithmeticException` or checking null instead of catching `NullPointerException`. I didn't check this logic yet, but `IllegalArgumentException` should only be used for validating quite straight-forward pre-conditions/argument checking, i.e. thrown for obvious bad input. If this is more than that (more complex logic), perhaps another exception type should be used instead, which could then be handled by caller. I'll have a closer look.

Reply:
Okay, checked all callers of `getApiUriBuilder()` and `getAppUriBuilder()`, and it seems we are free of bugs. :) If such a bug would be introduced, it would probably be detected immediately when testing the functionality, since basically the binding would stop working. Only thing to consider would be Javadoc for getApiUriBuilder() documenting that params takes pairs of parameter name and value.

Reply:
I checked all the calls to `appendParams` and an `IllegalArgumentException` will never happen.